### ISSUE TITLE
Compilation in sigle precision: issue when ccmake is used to configure compilation

### ISSUE THREAD

**Issue by lucafedeli88** [ECP-WarpX/WarpX#1781]
Hi! 
I've experienced a surprising issue concerning the compilation of WarpX in single precision.
If I do
```
mkdir build
cd build
cmake -DWarpX_PRECISION=SINGLE ../.
make -j8
```
everything works nicely.
However, if do:
```
mkdir build
cd build
ccmake ../.
make -j8
```
and if I configure  `WarpX_PRECISION=SINGLE` with `ccmake` I get the following error message (and the compilation fails):
```
[ 50%] Building CXX object CMakeFiles/WarpX.dir/Source/WarpX.cpp.o
/usr/bin/c++  -DBOOST_ALL_NO_LIB -DPXRMP_HAS_BOOST=1 -DWARPX_DIM_3D -DWARPX_PARSER_DEPTH=24 -DWARPX_QED -I/home/luca/Projects/warpx_directory/WarpX/Source -I/home/luca/Projects/warpx_directory/WarpX/build/_deps/fetchedamrex-src/Tools/C_scripts -I/home/luca/Projects/warpx_directory/WarpX/build/_deps/fetchedpicsar-src/multi_physics/QED/include -isystem /home/luca/Projects/warpx_directory/WarpX/build/_deps/fetchedamrex-src/Src/Base -isystem /home/luca/Projects/warpx_directory/WarpX/build/_deps/fetchedamrex-src/Src/Boundary -isystem /home/luca/Projects/warpx_directory/WarpX/build/_deps/fetchedamrex-src/Src/AmrCore -isystem /home/luca/Projects/warpx_directory/WarpX/build/_deps/fetchedamrex-src/Src/LinearSolvers/MLMG -isystem /home/luca/Projects/warpx_directory/WarpX/build/_deps/fetchedamrex-src/Src/LinearSolvers/Projections -isystem /home/luca/Projects/warpx_directory/WarpX/build/_deps/fetchedamrex-src/Src/Particle -isystem /home/luca/Projects/warpx_directory/WarpX/build/_deps/fetchedamrex-build -isystem /usr/include/x86_64-linux-gnu/mpich  -Wall -Wextra -Wpedantic -Wshadow -Woverloaded-virtual -Wunreachable-code -O3 -g -DNDEBUG   -pthread -fopenmp -o CMakeFiles/WarpX.dir/Source/WarpX.cpp.o -c /home/luca/Projects/warpx_directory/WarpX/Source/WarpX.cpp
In file included from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/WarpXParticleContainer.H:21,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/MultiParticleContainer.H:15,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/WarpX.H:16,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/WarpX.cpp:12:
/home/luca/Projects/warpx_directory/WarpX/Source/Particles/ElementaryProcess/QEDInternals/BreitWheelerEngineWrapper.H: In member function ‘int BreitWheelerEvolveOpticalDepth::operator()(amrex::ParticleReal, amrex::ParticleReal, amrex::ParticleReal, amrex::ParticleReal, amrex::ParticleReal, amrex::ParticleReal, amrex::ParticleReal, amrex::ParticleReal, amrex::ParticleReal, amrex::Real, amrex::ParticleReal&) const’:
/home/luca/Projects/warpx_directory/WarpX/Source/Particles/ElementaryProcess/QEDInternals/BreitWheelerEngineWrapper.H:163:39: error: cannot bind non-const lvalue reference of type ‘float&’ to an rvalue of type ‘float’
  163 |                 energy, chi_phot, dt, opt_depth, m_table_view);
      |                                       ^~~~~~~~~
In file included from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/ElementaryProcess/QEDInternals/BreitWheelerEngineWrapper.H:19,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/WarpXParticleContainer.H:21,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/MultiParticleContainer.H:15,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/WarpX.H:16,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/WarpX.cpp:12:
/home/luca/Projects/warpx_directory/WarpX/build/_deps/fetchedpicsar-src/multi_physics/QED/include/picsar_qed/physics/breit_wheeler/breit_wheeler_engine_core.hpp:134:40: note:   initializing argument 4 of ‘bool picsar::multi_physics::phys::breit_wheeler::evolve_optical_depth(RealType, RealType, RealType, RealType&, const TableType&, RealType) [with RealType = float; TableType = picsar::multi_physics::phys::breit_wheeler::dndt_lookup_table<float, picsar::multi_physics::containers::picsar_span<const float> >; picsar::multi_physics::phys::unit_system UnitSystem = picsar::multi_physics::phys::unit_system::SI]’
  134 |         const RealType t_dt, RealType& optical_depth,
      |                              ~~~~~~~~~~^~~~~~~~~~~~~
In file included from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/ElementaryProcess/Ionization.H:13,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/PhysicalParticleContainer.H:16,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/MultiParticleContainer.H:16,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/WarpX.H:16,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/WarpX.cpp:12:
/home/luca/Projects/warpx_directory/WarpX/Source/Particles/Gather/GetExternalFields.H: In member function ‘void GetExternalField::operator()(long int, amrex::ParticleReal&, amrex::ParticleReal&, amrex::ParticleReal&) const’:
/home/luca/Projects/warpx_directory/WarpX/Source/Particles/Gather/GetExternalFields.H:44:59: error: no match for call to ‘(const HostDeviceParser<4>) (amrex::ParticleReal&, amrex::ParticleReal&, amrex::ParticleReal&, const Real&)’
   44 |             field_x += m_xfield_partparser(x, y, z, m_time);
      |                                                           ^
In file included from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/WarpXParticleContainer.H:13,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/MultiParticleContainer.H:15,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/WarpX.H:16,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/WarpX.cpp:12:
/home/luca/Projects/warpx_directory/WarpX/Source/Parser/WarpXParserWrapper.H:31:5: note: candidate: ‘template<class ... Ts> std::enable_if_t<((sizeof... (Ts) == N) && amrex::Same<amrex::Real, Ts ...>::value), float> HostDeviceParser<N>::operator()(Ts ...) const [with Ts = {Ts ...}; int N = 4]’
   31 |     operator() (Ts... var) const noexcept
      |     ^~~~~~~~
/home/luca/Projects/warpx_directory/WarpX/Source/Parser/WarpXParserWrapper.H:31:5: note:   template argument deduction/substitution failed:
In file included from /usr/include/c++/10/bits/move.h:57,
                 from /usr/include/c++/10/bits/stl_pair.h:59,
                 from /usr/include/c++/10/utility:70,
                 from /usr/include/c++/10/array:38,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/Parser/WarpXParser.H:10,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/Utils/WarpXUtil.H:11,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/MultiParticleContainer.H:14,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/WarpX.H:16,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/WarpX.cpp:12:
/usr/include/c++/10/type_traits: In substitution of ‘template<bool _Cond, class _Tp> using enable_if_t = typename std::enable_if::type [with bool _Cond = false; _Tp = float]’:
/home/luca/Projects/warpx_directory/WarpX/Source/Parser/WarpXParserWrapper.H:31:5:   required by substitution of ‘template<class ... Ts> std::enable_if_t<((sizeof... (Ts) == 4) && amrex::Same<amrex::Real, Ts ...>::value), float> HostDeviceParser<4>::operator()<Ts ...>(Ts ...) const [with Ts = {double, double, double, float}]’
/home/luca/Projects/warpx_directory/WarpX/Source/Particles/Gather/GetExternalFields.H:44:59:   required from here
/usr/include/c++/10/type_traits:2554:11: error: no type named ‘type’ in ‘struct std::enable_if<false, float>’
 2554 |     using enable_if_t = typename enable_if<_Cond, _Tp>::type;
      |           ^~~~~~~~~~~
In file included from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/ElementaryProcess/Ionization.H:13,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/PhysicalParticleContainer.H:16,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/MultiParticleContainer.H:16,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/WarpX.H:16,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/WarpX.cpp:12:
/home/luca/Projects/warpx_directory/WarpX/Source/Particles/Gather/GetExternalFields.H:45:59: error: no match for call to ‘(const HostDeviceParser<4>) (amrex::ParticleReal&, amrex::ParticleReal&, amrex::ParticleReal&, const Real&)’
   45 |             field_y += m_yfield_partparser(x, y, z, m_time);
      |                                                           ^
In file included from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/WarpXParticleContainer.H:13,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/MultiParticleContainer.H:15,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/WarpX.H:16,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/WarpX.cpp:12:
/home/luca/Projects/warpx_directory/WarpX/Source/Parser/WarpXParserWrapper.H:31:5: note: candidate: ‘template<class ... Ts> std::enable_if_t<((sizeof... (Ts) == N) && amrex::Same<amrex::Real, Ts ...>::value), float> HostDeviceParser<N>::operator()(Ts ...) const [with Ts = {Ts ...}; int N = 4]’
   31 |     operator() (Ts... var) const noexcept
      |     ^~~~~~~~
/home/luca/Projects/warpx_directory/WarpX/Source/Parser/WarpXParserWrapper.H:31:5: note:   template argument deduction/substitution failed:
In file included from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/ElementaryProcess/Ionization.H:13,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/PhysicalParticleContainer.H:16,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/MultiParticleContainer.H:16,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/WarpX.H:16,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/WarpX.cpp:12:
/home/luca/Projects/warpx_directory/WarpX/Source/Particles/Gather/GetExternalFields.H:46:59: error: no match for call to ‘(const HostDeviceParser<4>) (amrex::ParticleReal&, amrex::ParticleReal&, amrex::ParticleReal&, const Real&)’
   46 |             field_z += m_zfield_partparser(x, y, z, m_time);
      |                                                           ^
In file included from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/WarpXParticleContainer.H:13,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/Particles/MultiParticleContainer.H:15,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/WarpX.H:16,
                 from /home/luca/Projects/warpx_directory/WarpX/Source/WarpX.cpp:12:
/home/luca/Projects/warpx_directory/WarpX/Source/Parser/WarpXParserWrapper.H:31:5: note: candidate: ‘template<class ... Ts> std::enable_if_t<((sizeof... (Ts) == N) && amrex::Same<amrex::Real, Ts ...>::value), float> HostDeviceParser<N>::operator()(Ts ...) const [with Ts = {Ts ...}; int N = 4]’
   31 |     operator() (Ts... var) const noexcept
      |     ^~~~~~~~
/home/luca/Projects/warpx_directory/WarpX/Source/Parser/WarpXParserWrapper.H:31:5: note:   template argument deduction/substitution failed:
make[2]: *** [CMakeFiles/WarpX.dir/build.make:63: CMakeFiles/WarpX.dir/Source/WarpX.cpp.o] Errore 1
make[2]: uscita dalla directory «/home/luca/Projects/warpx_directory/WarpX/build»
make[1]: *** [CMakeFiles/Makefile2:848: CMakeFiles/WarpX.dir/all] Errore 2
make[1]: uscita dalla directory «/home/luca/Projects/warpx_directory/WarpX/build»
make: *** [Makefile:130: all] Errore 2

```
If it can help, my `cmake` version is `CMake Version 3.16.3` and my compiler is `g++ (Ubuntu 10.2.0-13ubuntu1) 10.2.0`, but I have the same issue with `clang`. I am using the latest version of WarpX ( commit `b160cdd66d0ffe679d9e4755cd98e399d8ebc5dd` )

**Comment by ax3l** [comment]
Ok, what you describe should be identical to
```bash
rm -rf build
cmake -S . -B build
cmake -S . -B build -DWarpX_PRECISION=SINGLE
cmake --build build
```
and that workflow shows the same problem to me.

We probably need to trigger a clean rebuild of AMReX and/or PICSAR when this option changes in an existing `CMakeCache.txt`.

(The work-around until then is to remove `CMakeCache.txt` or simply clean the whole build directory.)

**Comment by ax3l** [comment]
Found the issue, the `AMReX_PARTICLES_PRECISION:STRING=DOUBLE` is not updated in `CMakeCache.txt`  after the 2nd call :). This is because it was renamed at some point. Fix proposed in #1782
